### DISCUSSION TITLE
Any interest in "geoh5vista"?

### DISCUSSION THREAD

**Discussion by derek-kinakin** [pyvista/pyvista#6064] [Show and tell]
The "omfvista" library provides a useful way to access OMF projects/data via PyVista. Starting from that example and structure, I've been prototyping an equivalent "geoh5vista" library that works with [Geoh5 files](https://github.com/MiraGeoscience/geoh5py).

I'm uploading my work-in-progress to see if folks have any interest in this kind of package:
[geoh5vista] (https://github.com/derek-kinakin/geoh5vista)

So far, I have got it to a stage where I can run the equivalent OMF reading/visualization examples on Geoh5 files.

There is plently left to do...
* Add support for rotated block models
* Add support for "referenced" data
* Finish the nested group loading support
* Add more geoh5 objects
* Turn it into a real package....
...and likely a lot more...

I'm keen to hear feedback on the usefulness of this project.

### Pseudocode or Screenshots

![image](https://github.com/pyvista/pyvista/assets/89031238/ae9a9b4e-c5f6-462b-ae74-cd9e2c4c3eb9)

**Comment by tkoyama010** [comment]
Awesome job! If you want to maintain this as a library, please consult with us here and we can advise you.

**Reply by derek-kinakin** [reply]
Thanks very much. I'll certainly be asking for advice!

**Reply by banesullivan** [reply]
Fantastic work!! I love seeing these sorts of interoperability layers between PyVista and domain-specific formats!

**Reply by derek-kinakin** [reply]
Thank you for 'omfvista'; especially considering how much I'm using it as a reference and jumping-off point.

**Comment by RichardScottOZ** [comment]
Sounds great - been doing xarray to geoh5.

**Comment by derek-kinakin** [comment]
Hi all, I've been slowly working away on this library and have recently made a number of edits and improvements. The code now supports reading most of the key entities from geoh5 files. I'm certainly keen for testers, contibutors, and feedback. The addition of the "write to geoh5" functions is in progress.

**Comment by derek-kinakin** [comment]
I've **finally** published this package:
https://pypi.org/project/geoh5vista/

Lot's of optimization and clean-up left to do. There are a few design decisions that it would be great to get feedback on:

1. Use of field_data vs user_dict for attributes/metadata (i.e. name, color) brought over from the geoh5 file. I've tried both options...I'm not sure if one is technically better than the other.
2. Drillhole handling. I've gone back and forth between a combined PolyData object and a Multiblock of PolyData objects.

I hope people give it a try and let me know about things that are not working/are broken.

**Reply by user27182** [reply]
I am not familiar with the file format, but my two cents as a PyVista maintainer (and author of the `user_dict` feature):
> 1. Use of field_data vs user_dict for attributes/metadata (i.e. name, color) brought over from the geoh5 file. I've tried both options...I'm not sure if one is technically better than the other.

If the metadata contains basic strings or numbers (i.e. scalar values), then go with user dict. Field data is annoying to use directly with scalars since everything must be a proper array, requiring users to index into the zeroth entry to get the actual data they want. But if the metadata can also contain proper arrays where the dtype matters, then serializing/deserializing as JSON for the user dict gets complicated, and so in this case proper arrays should _not_ be included in the user dict, but instead added as field data directly.

Also keep in mind that the user dict _is_ field data, but it's also a PyVista construct, so for interoperability with other libraries, these libraries would need to understand that there is a `_PYVISTA_USER_DICT` field data array with a JSON-serialized string array. If only PyVista is used to read these fields, then that's not an issue.

> 2. Drillhole handling. I've gone back and forth between a combined PolyData object and a Multiblock of PolyData objects.

If there is value in having separate PolyData, and the data typically has multiple objects, it's easy enough to do `pv.merge(multiblock)` to get a single PolyData out of it if desired. So in this case I would advocate for storing as MultiBlock, especially if the blocks have meaningful names, or their own separate metadata.

**Reply by derek-kinakin** [reply]
Thank you so much for the response and discussion.

1. user_dict vs field data

The values are typically single strings (object names), single booleans (visibility), or lists (color as RGB values). A basic use case for this info would be reading the geoh5 file and being able to quickly plot the data using PyVista with the same colors and names (legend entries) as used in the geoh5 file.

It seems that using the user_dict might be the way to go.

2. Drillholes as MultiBlock or combined PolyData

I'll look at the MultiBlock option again. The Polydata option made it simple to apply a tube filter for visualizing the drillholes better. If I can do the same with the MultiBlock, then it is an easy decision.

Time to edit some code!
